### ISSUE TITLE
Rule proposal: enforce variance annotations

### ISSUE THREAD

**Issue by DetachHead** [typescript-eslint/typescript-eslint#4799]
### Description

optional variance annotations were added in typescript 4.7 - https://devblogs.microsoft.com/typescript/announcing-typescript-4-7-beta/#optional-variance-annotations-for-type-parameters

it would be nice to have a rule that enforces them

### Fail

```ts
interface Foo<T> {
    value: T
}
```

### Pass

```ts
interface Foo<out T> {
    value: T
}
```

**Comment by bradzacher** [comment]
@DetachHead
TS has an automatic variance system, meaning 99.99% of the time a variance annotation is not needed and changes nothing. The annotations are really being added to help with some edge cases.

Why exactly would you want to manually annotate the variance every single time? What benefit would you gain from the verbosity?

**Comment by DetachHead** [comment]
there are several benefits that were mentioned in that blog post, such as:
- it's better for readability when the variance is explicit (same can be argued for existing rules such as [`explicit-function-return-type`](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/explicit-function-return-type.md))
- improves compiler performance because it doesn't need to perform expensive calculations to determine the variance
